### ISSUE TITLE
invalid function helm-build-sync-source

### ISSUE THREAD

**Issue by ar1a** [jacktasia/dumb-jump#224]
```
Debugger entered--Lisp error: (invalid-function helm-build-sync-source)
  helm-build-sync-source("Jump to: " :candidates ("+magit.el:77: (pretty-magit \"Docs\" ? '(:foreground \"#3F681C\" :height 1.0 :family \"FontAwesome\"))" "+magit.el:76: (pretty-magit \"Clean\" ? '(:foreground \"#FFBB00\" :height 1.0 :family \"FontAwesome\"))" "+magit.el:75: (pretty-magit \"Fix\" ? '(:foreground \"#FB6542\" :height 1.0 :family \"FontAwesome\"))" "+magit.el:74: (pretty-magit \"Add\" ? '(:foreground \"#375E97\" :height 1.0 :family \"FontAwesome\"))" "+magit.el:73: (pretty-magit \"Feature\" ? '(:foreground \"slate gray\" :height 1.0 :family \"FontAwesome\"))" "+magit.el:69: (def-package! pretty-magit" "+magit.el:79: (pretty-magit \"origin\" ? '(:box t :height 1.0 :family \"octicons\") t))" "+magit.el:78: (pretty-magit \"master\" ? '(:box t :height 1.0 :family \"octicons\") t)" "local/pretty-magit.el:26: \"Add face properties and compose symbols for buffer from pretty-magit.\"" "local/pretty-magit.el:15: (defmacro pretty-magit (WORD ICON PROPS &optional NO-PROMPT?)" "local/pretty-magit.el:7: (provide 'pretty-magit)") :persistent-action dumb-jump-helm-persist-action)
  dumb-jump-prompt-user-for-choice("/home/aria/.doom.d" ((:path "+magit.el" :line 77 :context "  (pretty-magit \"Docs\" ? '(:foreground \"#3F681C\" :height 1.0 :family \"FontAwesome\"))" :diff 1 :target "pretty-magit") (:path "+magit.el" :line 76 :context "  (pretty-magit \"Clean\" ? '(:foreground \"#FFBB00\" :height 1.0 :family \"FontAwesome\"))" :diff 2 :target "pretty-magit") (:path "+magit.el" :line 75 :context "  (pretty-magit \"Fix\" ? '(:foreground \"#FB6542\" :height 1.0 :family \"FontAwesome\"))" :diff 3 :target "pretty-magit") (:path "+magit.el" :line 74 :context "  (pretty-magit \"Add\" ? '(:foreground \"#375E97\" :height 1.0 :family \"FontAwesome\"))" :diff 4 :target "pretty-magit") (:path "+magit.el" :line 73 :context "  (pretty-magit \"Feature\" ? '(:foreground \"slate gray\" :height 1.0 :family \"FontAwesome\"))" :diff 5 :target "pretty-magit") (:path "+magit.el" :line 69 :context "(def-package! pretty-magit" :diff 9 :target "pretty-magit") (:path "+magit.el" :line 79 :context "  (pretty-magit \"origin\" ? '(:box t :height 1.0 :family \"octicons\") t))" :diff -1 :target "pretty-magit") (:path "+magit.el" :line 78 :context "  (pretty-magit \"master\" ? '(:box t :height 1.0 :family \"octicons\") t)" :diff 0 :target "pretty-magit") (:path "local/pretty-magit.el" :line 26 :context "  \"Add face properties and compose symbols for buffer from pretty-magit.\"" :diff 52 :target "pretty-magit") (:path "local/pretty-magit.el" :line 15 :context "(defmacro pretty-magit (WORD ICON PROPS &optional NO-PROMPT?)" :diff 63 :target "pretty-magit") (:path "local/pretty-magit.el" :line 7 :context "(provide 'pretty-magit)" :diff 71 :target "pretty-magit")))
```

Strange, because the function isn't invalid

**Comment by jacktasia** [comment]
Thanks for the report! A similar thing was reported in #153 can you please check that out and see if it works for you? Also, a fix for helm was just merged in yesterday via #222 so updating dumb-jump too might help.

Please let me know if you continue have problems after trying those things and I will try to dig in more. Thanks!

**Comment by ar1a** [comment]
Yeah, I upgraded all my packages and it's still happening.

**Comment by jacktasia** [comment]
> Yeah, I upgraded all my packages and it's still happening.

What version of Emacs are you using? What OS?

**Comment by ar1a** [comment]
26.1, emacs.

My config can be found https://github.com/ar1a/dotfiles/tree/master/emacs/.doom.d which is a private config for https://github.com/hlissner/doom-emacs/tree/develop (note the develop branch)

**Comment by jacktasia** [comment]
Thanks for the info. I tried using doom-emacs (develop) with a fresh emacs 26.1 and dumb-jump worked as expected. I also tried using doom-emacs (master) with a fresh emacs 26.1 and dumb-jump also worked as expected. 

As far as I can tell this is an issue with your configuration and not dumb-jump. If you can provide a link to a diff between your config and doom-emacs/develop default then I can try to help from there. Addtionally you might try opening an issue with doom-emacs. It could be something like this: https://github.com/hlissner/doom-emacs/issues/227

**Comment by ar1a** [comment]
My config isnt an overlap of doom more as an extension. If you grab my config and chuck it in ~/.doom.d, and then go into ~/.emacs.d and run make, you'll have my environment (for the most part).

**Comment by jacktasia** [comment]
> My config isnt an overlap of doom more as an extension. If you grab my config and chuck it in ~/.doom.d, and then go into ~/.emacs.d and run make, you'll have my environment (for the most part).

I only went down the rabbit hole for a while but I am pretty sure you're running into this:

https://github.com/emacs-helm/helm/issues/602

Not sure how you're doing things differently than doom-emacs, but it clearly seems like an issue with your personal configuration (vs doom-emacs or dumb-jump)

I am going to close this since it doesn't seem like a bug in dumb-jump. Please re-open if you find out that is not the case. Thanks!

**Comment by hlissner** [comment]
After some testing, it appears that this is a load-order issue at compile-time. If you install helm _after_ dumb-jump (and  they are byte-compiled in that order by package-install), then `helm-build-sync-source` won't be expanded correctly and is treated as a function.

Would it not make sense to add a compile-time `require` to dumb-jump? e.g.

```elisp
(eval-when-compile
  (require 'helm-source nil t))
```

**Comment by jacktasia** [comment]
@hlissner Thanks for looking into this!

> By the way, would it not make sense to add a compile-time require to dumb-jump? e.g.

Maybe, this seems potentially problematic because helm isn't required for dumb-jump because you could use popup or ivy instead. Surely this problem has been solved before for packages that support either helm or ivy. I will do some more research, but I am very open to any other ideas or PRs. Thanks again!

**Comment by xuchunyang** [comment]
I encounter the same issue today when making my own package, @hlissner's analysis is correct, the byte compiled code should not contain macros otherwise Emacs will emit an Invalid function error, so I think we cam simply drop `helm-build-sync-source` and switch to `helm-make-source`, the macro is simple anyway:


```el
(defmacro helm-build-sync-source (name &rest args)
  "Build a synchronous helm source with name NAME.
Args ARGS are keywords provided by `helm-source-sync'."
  (declare (indent 1))
  `(helm-make-source ,name 'helm-source-sync ,@args))
```

**Comment by matthuszagh** [comment]
I agree with @xuchunyang here. I'm coming from Nix, where helm needs to be available at byte compilation in order for the current implementation to work. The alternative with `helm-make-source` is simple, as mentioned.

```elisp
(defun dumb-jump-prompt-user-for-choice (proj results)
    "Put a PROJ's list of RESULTS in a 'popup-menu' (or helm/ivy)
for user to select.  Filters PROJ path from files for display."
    (let ((choices (--map (dumb-jump--format-result proj it) results)))
      (cond
       ((and (eq dumb-jump-selector 'ivy) (fboundp 'ivy-read))
        (funcall dumb-jump-ivy-jump-to-selected-function results choices proj))
       ((and (eq dumb-jump-selector 'helm) (fboundp 'helm))
        (helm :sources
          (helm-make-source "Jump to: " 'helm-source-sync
            :action '(("Jump to match" . dumb-jump-result-follow))
            :candidates (-zip choices results)
            :persistent-action 'dumb-jump-helm-persist-action)
          :buffer "*helm dumb jump choices*"))
       (t
        (dumb-jump-to-selected results choices (popup-menu* choices))))))
```

**Comment by jacktasia** [comment]
@matthuszagh Thanks! Trying this in #315

**Comment by bhipple** [comment]
I've discovered this package today, and after spending a ton of time over the last couple years wrestling with ctags, GNU Global rebuild hooks, LSP backends that don't quite work or are really slow, `dumb-jump` is a breath of fresh air. However, when I turned on Helm I hit this issue; using your fix in #315 resolves it for me.

It'd be great if we could merge the bugfix to master!  While my setup is now working, it took me a while to debug, fork, and update `use-package` correctly.
